### PR TITLE
fix:Refactor movement logic for hidden unit handling

### PR DESCRIPTION
**Pull request by dangdinhbaohoang12**
- New Features:
- [x]  Units can now detect and respond to unseen units from any civilization, including friendly and allied forces.
- [x] Targeting an unseen hostile unit reveals it, consumes movement, and issues a war notification without entering its tile.
- Bug Fixes:
- [x] Transported units now follow the carrier’s actual movement path, avoiding hidden blocked tiles.
- [x] Strict placement checks continue to prevent teleportation and unit swapping onto occupied tiles.
- [x] Units now detect hidden opposing military units before entering blocked tiles.
- [x] Discovering a hidden military unit no longer consumes movement or causes entry or overwriting.
- [x] Movement options and previews update correctly after discovery.
- [x] Routes remain viable through tiles blocked only by undetected enemy military units and reroute correctly afterward.
- [x] Hidden enemy civilians remain enterable and capturable.
Discovered hidden-unit locations remain accurate after visibility updates, including when units move or are destroyed.
Fixing the issue—partially or completely? I honestly don't know. :) :https://github.com/yairm210/Unciv/issues/7670

### CODE REVIEW

**Comment by dangdinhbaohoang12** [comment]
Fixed all the issues present in the old pull requests: https://github.com/yairm210/Unciv/pull/15530

File: core/src/com/unciv/logic/map/mapunit/movement/UnitMovement.kt [hunk 1]
```diff
@@ -560,6 +573,18 @@ class UnitMovement(val unit: MapUnit) {
             if (!unit.civ.gameInfo.gameParameters.godMode)
                 passingMovementSpent += MovementCost.getMovementCostBetweenAdjacentTiles(unit, previousTile, tile)
 
+            // We were allowed to *attempt* moving into this tile even though it secretly contains
+            // a unit of another civ we couldn't see (friendly/allied or enemy, e.g. an undetected
+            // submarine) - moving into it is what reveals that unit, so we stop here instead of
+            // actually entering the tile and overwriting/stacking onto it.
+            val hiddenBlocker = getHiddenBlockingUnit(tile)
+            if (hiddenBlocker != null) {
+                unit.useMovementPoints(unit.currentMovement)
```
Comment (yairm210) [inline core/src/com/unciv/logic/map/mapunit/movement/UnitMovement.kt]: Do we use up all the movement points? Shouldn't we only use the movement required to get to where we got? In which case this is passingMovementSpent

**Comment by yairm210** [comment]
Looks pretty good
There is still nothing that makes the "revealed unit" actually visible though, right? Just validating

**Comment by dangdinhbaohoang12** [comment]
That's just the previous version. I already have this one:
unit.civ.viewableInvisibleUnitsTiles =
    unit.civ.viewableInvisibleUnitsTiles + tile

When `MapUnit.isVisibleTo(civ)` checks an invisible unit, it allows it when its tile is in `viewableInvisibleUnitsTiles`. The map visualization uses the same visibility logic, so the discovered unit is actually shown, not just notified.

**Comment by dangdinhbaohoang12** [comment]
@yairm210  Can I go ahead and make the changes, or are there still issues? If there are any remaining issues, I would appreciate your feedback.

**Comment by dangdinhbaohoang12** [comment]
@yairm210 Note: Since I am using Google Translate, some of my sentences may be incorrect or incomplete. I apologize for any errors.

**Comment by dangdinhbaohoang12** [comment]
After `viewableInvisibleUnitsTiles` changes, I immediately clear the unit's old pathfinding cache and the escort's cache, preventing stale paths from being reused.

**Comment by yairm210** [comment]
Looking at the cannotMoveTo - where ideally, civilians check civilian position and military check military, and now we have a mix-up since there could be invisible enemy units - it's clear to me that we need a CannotMoveToReason.TileIsNotEmptyHiddenUnit result. 

Then, cannotPassThroughReason doesn't hide any data (whether you use the data depends on caller); canMoveTo will, if it gets that result, "defer" it to the end and return any other "cannot move to" reason first (e.g. city), and only at the end return "invisible unit"; And then "thinks it can move to" logic is "if reason we can't move to is not invisibility reason".

**Comment by yairm210** [comment]
Don't want to discourage, we're close :D
This is a complicated one :)

**Comment by dangdinhbaohoang12** [comment]
And yet, I thought you had dropped this request :)

**Comment by dangdinhbaohoang12** [comment]
> Looking at the cannotMoveTo - where ideally, civilians check civilian position and military check military, and now we have a mix-up since there could be invisible enemy units - it's clear to me that we need a CannotMoveToReason.TileIsNotEmptyHiddenUnit result.
> 
> Then, cannotPassThroughReason doesn't hide any data (whether you use the data depends on caller); canMoveTo will, if it gets that result, "defer" it to the end and return any other "cannot move to" reason first (e.g. city), and only at the end return "invisible unit"; And then "thinks it can move to" logic is "if reason we can't move to is not invisibility reason".

I forgot I was supposed to do it this way.

**Comment by dangdinhbaohoang12** [comment]
Here is a  new update information: I have partially fixed the issue and am currently running tests at https://github.com/dangdinhbaohoang12/Unciv/pull/8. However, there are still some edge cases—previously pointed out to me—that remain:

Movement of an escorted pair into a tile containing a hidden unit (I traced the logic and it appears correct, but I was unable to run `twoEscortsCanSwap`-style escort tests against it).
Air units/paratroopers intentionally still use the strict `canMoveTo` check for their direct-landing branch in `getTileToMoveToThisTurn`—please confirm if this is the desired behavior (I left it as-is, as it reads like intentional "placement" semantics).
Dual-slot occupancy (where both a civilian and a military unit are hidden simultaneously) is handled in `getCannotMoveToReason` but is not covered by my new test. 
I would appreciate it if you could review whether these points are acceptable as a minimum viable solution, given the complexity involved in the fix.

**Comment by dangdinhbaohoang12** [comment]
>Movement of an escorted pair into a tile containing a hidden unit (I traced the logic and it appears correct, but I was unable to run twoEscortsCanSwap-style escort tests against it).
Air units/paratroopers intentionally still use the strict canMoveTo check for their direct-landing branch in getTileToMoveToThisTurn—please confirm if this is the desired behavior (I left it as-is, as it reads like intentional "placement" semantics).
Dual-slot occupancy (where both a civilian and a military unit are hidden simultaneously) is handled in getCannotMoveToReason but is not covered by my new test.

I gave this to Claude to test, and that’s what it said :). It really is a tricky issue.

**Comment by dangdinhbaohoang12** [comment]
@yairm210 can you go this and see file change:https://github.com/dangdinhbaohoang12/Unciv/pull/8.. and see if there are any other changes you'd like me to make right away. I'll make the edits and post them here.

**Comment by dangdinhbaohoang12** [comment]
@yairm210  Have you checked the changes in there yet?

**Comment by dangdinhbaohoang12** [comment]
@yairm210  In your opinion, has this pull request completely fixed the issue in this:https://github.com/yairm210/Unciv/issues/7670

File: core/src/com/unciv/logic/civilization/transients/CivInfoTransientCache.kt [hunk 2]
```diff
@@ -26,6 +27,8 @@ import com.unciv.logic.automation.Timers.Companion.timeThis
 /** CivInfo class was getting too crowded */
 class CivInfoTransientCache(val civInfo: Civilization) {
 
+    private val discoveredInvisibleUnitTiles = HashMap<MapUnit, Tile>()
```
Comment (yairm210) [inline core/src/com/unciv/logic/civilization/transients/CivInfoTransientCache.kt:30]: Doesn't belong in the transient cache, since if e.g. we save/load the game we want it to remain. This should in in Civilization.kt, and copied on clone()

Comment (yairm210) [reply core/src/com/unciv/logic/civilization/transients/CivInfoTransientCache.kt:30]: Also it should be serializable - that is, should have unit ID and tile position

**Comment by dangdinhbaohoang12** [comment]
I think this pull request needs to be fixed and merged quickly; I'm really tired of dealing with merge conflicts. :(

**Comment by dangdinhbaohoang12** [comment]
@yairm210 In your opinion, has this pull request completely fixed the issue in this:https://github.com/yairm210/Unciv/issues/7670

**Comment by dangdinhbaohoang12** [comment]
>I think this pull request needs to be fixed and merged quickly; I'm really tired of dealing with merge conflicts. :(

But I am still willing to fix this code if there are any errors. and problems.

**Comment by dangdinhbaohoang12** [comment]
@yairm210 In your opinion, has this pull request completely fixed the issue in this:https://github.com/yairm210/Unciv/issues/7670

**Comment by dangdinhbaohoang12** [comment]
If there is anything wrong with the code, I’d appreciate your feedback. :)

**Comment by dangdinhbaohoang12** [comment]
Is there anything left to fix regarding this issue, @yairm210 ?

**Comment by dangdinhbaohoang12** [comment]
I've fixed the errors (as far as I can see), so please review the code. I'm all set. :)

**Comment by dangdinhbaohoang12** [comment]
@SomeTroglodyte It seems like Yairn is busy today, right?

**Comment by dangdinhbaohoang12** [comment]
@yairm210 In your opinion, has this pull request completely fixed the issue in this:https://github.com/yairm210/Unciv/issues/7670

**Comment by dangdinhbaohoang12** [comment]
@yairm210  
in a new commit:
Bug Fixes
-Corrected notifications when units discover hidden blockers during movement.
-Discovering a hidden unit from a civilization at war now generates a War notification.
-Discovering a hidden unit from a civilization with another diplomatic relationship now generates a neutral Units notification identifying its civilization.
-Movement pathfinding and discovery behavior remain unchanged.

**Comment by dangdinhbaohoang12** [comment]
@yairm210  i need you review to complete this pull request before it's make conflicts. :)
